_posts/2023-06-27-redis.md

```diff
@@ -0,0 +1,23 @@
+---
+layout: post
+title: "Redis"
+description: "Why is redis fast, what are the pros and cons of using redis"
+comments: true
+keywords: "redis, infrastructure"
+published: false
+---
+
+[Redis](https://redis.io) wass designed to be a blazingly fast key-value store to make caching easy. And for most parts it lives up to the billing. Setting up and using redis has always been a breeze. The success and popularity of redis has makes it it a very powerful tool in the software engineers arsenal, with its support of almost all data type and associated operations out of the box, it is a perfect solution for all persistence problems; almost. As with everything, people have used it in ways without consideration for if its fitness for the problem at hand often to great success.
+
+Redis is famed for being able to perform operations that is measured in sub milliseconds. And the key to this is how it is designed. 
+It is designed to store data in memory.  If your first year computer science class is anything to go by, you will know this provides much fast read and write throughput  compared to disk by miles. When we create data structures using variables we often use arrays, hashes and maps. Redis supports these and more. it is also effecient and manipulating them because, It takes advantage of the underlying data structures that are optimized for in memory storage without worrying about how to persist it to durable storage. .It is also single threaded, Though a performant single thread system might sound counter intuitive, that are some peculiar performant advantages to it. And redis takes advantage of this in a brilliant way to ensure consistency without an cost to performance. Redis's single-thread will scales indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like CPU might become a bottle neck with this design but it turns out more often than not, you will hit a network bottleneck well before CPU cannot keep up. The positive side effect of this design is that atomicity of all operations come at no extra cost. Redis also use a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/), that is much more terse.  Couple the isolated event loop with a proprietary communication protocol and you have a blazing fast in memory data store that scales indefinitely in theory. 
```

Comment:
```suggestion
It is designed to store data in memory.  If your first year computer science class is anything to go by, you will know this provides much fast read and write throughput  compared to disk by miles. When we create data structures using variables we often use arrays, hashes and maps. Redis supports these and more. it is also effecient and manipulating them because, It takes advantage of the underlying data structures that are optimized for in memory storage without worrying about how to persist it to durable storage. It is also single threaded, Though a performant single thread system might sound counter intuitive, that are some peculiar performant advantages to it. And redis takes advantage of this in a brilliant way to ensure consistency without any cost to performance. Redis's single-thread will scales indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like CPU might become a bottle neck with this design but it turns out more often than not, you will hit a network bottleneck well before CPU cannot keep up. The positive side effect of this design is that atomicity of all operations come at no extra cost. Redis also use a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/), that is much more terse.  Couple the isolated event loop with a proprietary communication protocol and you have a blazing fast in memory data store that scales indefinitely in theory and you have a highly preformant database. 
```